designs/2018-async-commands/README.md

```diff
@@ -0,0 +1,81 @@
+
+## Detailed Design
+
+Based on discussion in the original [PR](https://github.com/eslint/eslint/issues/10606#issue-341171187)there are some main points.
+
+1. The features that are intended to be async/parallel need to be behind a flag on the CLI, and opt in when using the API (maybe through an async command or optoin when using CLIEngine).
+2. There are two options here: `promise` based and `thread` based. Some users suggested looking at [esprint](https://github.com/pinterest/esprint) for inspiration. Some of the concerns were that `promises` would not be faster per say depending on the implementation. On the flipside, it has been noted that `threads`/workers can be expensive on platforms like Windows. Ideally, the implementation would use a balance of both. For example, `jest` has async methods that use threads under the hood.
+3. The experience should be opt in so as not to break existing users.
+4. There should be some sort of mechanism to handle eslint rules that expect `sync` reading. Rules may need to identify themselves as a "type". This way, if a rule is being loaded that requires `sync`, the code can handle it accordingly. Any rule that doesn't identify as a `sync` or `async` would be defaulted to `sync`.
+
+## Documentation
+
+1. There should be JSDoc code for sure. In code documentation is always the best.
+2. I do believe this should warrant a blog announcement to say "hey we are moving in this direction, and would like feedback and help with it". Many companies would contribute and help once the initial implementation is there.
+
+## Drawbacks
+
+1. More to maintain for the maintainers
+2. Threading and async problems are hard, and can require considerable effort (even with other projects out there doing it).
+3. This may expose unintended edge cases for eslint rules that expect synchronous scanning. There will need to be a shim mechanism for sure.
+
+## Backwards Compatibility Analysis
+
+As long as this functionality is marked as "beta" until ready and hid behind a flag or options, it should not break any existing users.
+
+## Alternatives
+
+https://github.com/pinterest/esprint <--- The main option, but Pinterest seems to have lost interest in it
+
+I tried wrapping ESLint in promises, but that was very messy and didn't work out too well.
+I tried making a PR that would change the functions I cared about, but the general discussion (in the PR above) showed there would need to be a more thoughtful attempt.
+
+## Open Questions
+
+1. What is the core architecture of ESLint that requires it to be synchronous?
+
+## Help Needed
+
+<!--
+    This section is optional.
+
+    Are you able to implement this RFC on your own? If not, what kind
+    of help would you need from the team?
+-->
+I would like help from the core team with this effort as I do work full time and even this has been a few months of work in my free time outside of work.
```

Comment:
What type of help are you talking about? Editing the RFC? Implementing the change?

Reply:
Implementing. I tend to only have weekends as time when I can work on side projects.  Otherwise, I don't mind taking a few months on weekends working on this 😄. It will be a lot of code reviews for sure is all.